include/oneapi/dpl/pstl/hetero/numeric_impl_hetero.h

```diff
@@ -108,23 +118,244 @@ __pattern_transform_reduce(_ExecutionPolicy&& __exec, _ForwardIterator __first,
     return __res;
 }
 
+template<bool _Inclusive, ::std::uint16_t _ElemsPerItem = 0, ::std::uint16_t _WGSize = 0, bool _IsFullGroup = false>
+struct __single_group_scan
+{
+    template<
+        typename _KernelName, typename _Policy, typename _InRng, typename _OutRng, typename _InitType, typename _BinaryOperation, typename _UnaryOp
+#if _ONEDPL_COMPILE_KERNEL
+        , typename _Kernel
+#endif
+    >
+    static void __launch_dynamic_bounds_scan(const _Policy& __policy, _InRng __in, _OutRng __out, ::std::size_t __n, _InitType __init, _BinaryOperation __bin_op, _UnaryOp __unary_op, ::std::uint16_t __wg_size
+#if _ONEDPL_COMPILE_KERNEL
+            , _Kernel&& __kernel
+#endif
+            )
+    {
+        using _ValueType = typename _InitType::__value_type;
+
+        const ::std::uint16_t __elems_per_item = __par_backend_hetero::__ceiling_div(__n, __wg_size);
+        const ::std::uint16_t __elems_per_wg = __elems_per_item * __wg_size;
+
+        auto __event = __policy.queue().submit([&](sycl::handler& __hdl) {
+            oneapi::dpl::__ranges::__require_access(__hdl, __in, __out);
+
+            auto __lacc =  __dpl_sycl::__local_accessor<_ValueType>(sycl::range<1>{__elems_per_wg}, __hdl);
+#if _ONEDPL_COMPILE_KERNEL && _ONEDPL_KERNEL_BUNDLE_PRESENT
+            __hdl.use_kernel_bundle(__kernel.get_kernel_bundle());
+#endif
+            __hdl.parallel_for<_KernelName>(
+#if _ONEDPL_COMPILE_KERNEL && !_ONEDPL_KERNEL_BUNDLE_PRESENT
+                __kernel,
+#endif
+                sycl::nd_range<1>(__wg_size, __wg_size), [=](sycl::nd_item<1> __self_item) {
+                const auto& __group = __self_item.get_group();
+                // This kernel is only launched for sizes less than 2^16
+                const ::std::uint16_t __item_id = __self_item.get_local_linear_id();
+
+                for (::std::uint16_t __idx = __item_id; __idx < __n; __idx += __wg_size)
+                {
+                   __lacc[__idx] = __unary_op(__in[__idx]);
+                }
+
+                __group_scan<_ValueType>(__group, __lacc.get_pointer(), __lacc.get_pointer() + __n, __bin_op, __init);
+
+                for (::std::uint16_t __idx = __item_id; __idx < __n; __idx += __wg_size)
+                {
+                   __out[__idx] = __lacc[__idx];
+                }
+
+                const ::std::uint16_t __residual = __n % __wg_size;
+                const ::std::uint16_t __residual_start = __n - __residual;
+                if (__residual > 0 && __item_id < __residual)
+                {
+                    auto __idx = __residual_start + __item_id;
+                   __out[__idx] = __lacc[__idx];
+                }
+            });
+        });
+        __event.wait();
```

Comment:
We don't use explicit synchronization for kernel pipeline, just kernels dependency, which defined by setting a depended event, for example
https://github.com/oneapi-src/oneDPL/blob/main/include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h#L455

Reply:
There was already a `wait()` at the end of the `transform_scan` implementation. See this code on main:  https://github.com/oneapi-src/oneDPL/blob/493b832a16643e4284a2085d99c68861ae5e7c4d/include/oneapi/dpl/pstl/hetero/numeric_impl_hetero.h#L143-L155

The `wait()` here is the same as the one that exists on main.

Reply:
1.  sycl kernels and submit ones should not be placed in that file `numeric_impl_hetero.h`. So, please move that code to "dpcpp/unseq_backend_sycl.h"
2 Moreover, The all sycl specific should be placed in `dpcpp` folder.

After that, looking on the `__parallel_scan_submitter` for example you will notice that there is now wait in the scan pattern of SYCL backend. The scan pattern of SYCL backend is used for async API as well.
https://github.com/oneapi-src/oneDPL/blob/main/include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h#L285